Added support for FunctionType::Kind::StringConcat (#12628)
ethereum/solidity: docs/types/reference-types.rst

    that you are accessing the low-level bytes of the UTF-8 representation,
    and not the individual characters.

> .. index:: ! bytes-concat

Reviewer: Please keep the index. You can also add multiple items here.

.. _string-concat:

> ``<bytes/string>.concat`` function

Reviewer: What about "The functions ``bytes.concat`` and ``string.concat``"

> The functions ``bytes-concat`` and ``string-concat``

Reviewer: Why not bytes.concat and string.concat?

> Using the string.concat function you can concatenate a variable number of ``string`` variables using ``string.concat``.

Reviewer: ```suggestion
You can concatenate a arbitrary number of ``string`` values using ``string.concat``.
```

> You can concatenate a arbitrary number of ``string`` values using ``string.concat``.

Reviewer: ```suggestion
You can concatenate an arbitrary number of ``string`` values using ``string.concat``.
```

> The function returns a single ``string memory`` array/array-slices that contains the contents of the arguments without padding.

Reviewer: The arguments can be strings or string slices, not the return value.

Reviewer: Maybe just add to the previous sencence: You can use anything that can be implicitly converted to `string memory`, which also includes string slices and string literals.

> Analogously, the bytes.concat function can concatenate a variable number of ``bytes`` or ``bytes1 ... bytes32`` using ``bytes.concat``.

Reviewer: ```suggestion
Analogously, the ``bytes.concat`` function can concatenate a variable number of ``bytes`` or ``bytes1 ... bytes32`` values.
```

Reviewer: Actually I think for both bytes.concat and string.concat you can also use array slices

> Analogously, the ``bytes.concat`` function can concatenate an arbitrary number of ``bytes`` or ``bytes1 ... bytes32`` values/array-slices.

Reviewer: ```suggestion
You can concatenate an arbitrary number of ``string`` values using ``string.concat``.
The function returns a single ``string memory`` array that contains the contents of the arguments without padding.
If you want to use parameters of other types that are not implicitly convertible to ``string``, you need to convert them to ``string`` first.

Analogously, the ``bytes.concat`` function can concatenate an arbitrary number of ``bytes`` or ``bytes1 ... bytes32`` values.
```

Since we also don't mention string literals, maybe adding "not implicitly convertible" is enough.

If you want to use parameters of other types, you need to convert them to ``string`` first.

You can concatenate a variable number of ``bytes`` or ``bytes1 ... bytes32`` using ``bytes.concat``.
Analogously, the bytes.concat function can concatenate a variable number of ``bytes`` or ``bytes1 ... bytes32`` using ``bytes.concat``.
The function returns a single ``bytes memory`` array that contains the contents of the arguments without padding.
If you want to use string parameters or other types, you need to convert them to ``bytes`` or ``bytes1``/.../``bytes32`` first.

.. code-block:: solidity

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity ^0.8.4;
    pragma solidity ^0.8.12;

    contract C {
        bytes s = "Storage";

>         function f(bytes calldata c, string memory m, bytes16 b) public view {

Reviewer: Can you have at least one `bytes16` (or other fixed bytes) variable in the example?

>         function f(bytes calldata c, string memory m, bytes16 b) public view {

Reviewer: Maybe call the variables `bc`, `sm` and `b`?

>             string memory concat_string = string.concat(s, string(c), "Literal", m , b);

Reviewer: ```suggestion
            string memory concat_string = string.concat(s, string(c), "Literal", m, b);
```

    }

> If you call ``string.concat`` without arguments it will return an empty ``string`` array.
> Whereas, If you call ``bytes.concat`` without arguments it will return an empty ``bytes`` array.

Reviewer: ```suggestion
If you call ``string.concat`` or ``bytes.concat`` without arguments they return an empty array.
```

    that you are accessing the low-level bytes of the UTF-8 representation,
    and not the individual characters.

.. index:: ! string-concat

.. _string-concat:

``string.concat`` function
^^^^^^^^^^^^^^^^^^^^^^^^^^

> You can concatenate a variable number of ``string`` variables using ``string.concat``.

Reviewer: Wouldn't it be better to only have one section for both?
